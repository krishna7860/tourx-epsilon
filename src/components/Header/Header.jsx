/* eslint-disable react/forbid-prop-types */
/* eslint-disable @typescript-eslint/no-shadow */
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import React, { useState, useEffect } from "react";
import {
  Toolbar,
  Drawer,
  MenuItem,
  Dialog,
  DialogContent,
} from "@material-ui/core";
import ExitToAppIcon from "@material-ui/icons/ExitToApp";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import MenuIcon from "@material-ui/icons/Menu";
import {
  HeaderWrapper,
  StyledToolbar,
  DrawerContainer,
  StyledAppBar,
  MenuButton,
  LogoItem,
  InnerItem,
  ActionButton,
  MenuIconButton,
  EndWrapper,
  RelativeWrapper,
  Profile,
  List,
  Avatar,
  Span,
} from "./Style";
import { setIsAuth, logoutUser } from "../../redux/commonActions/auth";
import Signup from "../../containers/WelcomeGuide/SignUpForm/Signup";
import Login from "../LoginForm/Login";
import About from "../About/About";

const headersData = [
  {
    label: "About us",
    href: "/",
  },
];
const Header = ({
  isTransparent,
  elevation,
  isAuth,
  setIsAuth,
  logoutUser,
  user,
}) => {
  const [state, setState] = useState({
    mobileView: false,
    drawerOpen: false,
    loading: false,
    open: false,
    loginOpen: false,
    profileOpen: false,
    aboutOpen: false,
  });

  const { mobileView, drawerOpen, open, loginOpen, profileOpen, aboutOpen } =
    state;

  useEffect(() => {
    const setResponsiveness = () => {
      return window.innerWidth < 900
        ? setState((prevState) => ({ ...prevState, mobileView: true }))
        : setState((prevState) => ({ ...prevState, mobileView: false }));
    };

    setResponsiveness();

    window.addEventListener("resize", () => setResponsiveness());

    return () => {
      window.removeEventListener("resize", () => setResponsiveness());
    };
  }, []);

  const logout = () => {
    setState({ ...state, loading: true });
    localStorage.removeItem("AuthToken");
    logoutUser();
    setTimeout(() => {
      setState({ ...state, loading: true });
      setIsAuth(false);
    }, 2000);
  };

  const handleProfileClose = (e) => {
    if (
      e.target.id !== "profile" &&
      e.target.id !== "my-profile" &&
      e.target.parentElement.id !== "my-profile"
    ) {
      setState({ ...state, profileOpen: false });

      document.getElementsByTagName("body")[0].removeEventListener("click");
    }
  };

  const handleProfileOpen = () => {
    setState({ ...state, profileOpen: true });

    document.getElementsByTagName("body")[0].addEventListener("click", (e) => {
      handleProfileClose(e);
    });
  };

  // Logo
  const tourxLogo = (
    <LogoItem to="/">
      <InnerItem isColored={isTransparent}>TourX</InnerItem>
    </LogoItem>
  );
  // Get Menus
  const getMenuButtons = () => {
    return (
      <>
        {headersData.map(({ label }) => {
          return (
            <MenuButton
              {...{
                key: label,
              }}
              isTransparent={isTransparent}
              onClick={() => setState({ ...state, aboutOpen: true })}
              key={label}
            >
              {label}
            </MenuButton>
          );
        })}
        {isAuth ? (
          <>
            <ActionButton
              isTransparent={isTransparent}
              style={{ background: "transparent" }}
              id="my-profile"
              onClick={handleProfileOpen}
            >
              My Account
            </ActionButton>
            <RelativeWrapper>
              {profileOpen ? (
                <Profile id="profile">
                  <List className="profile" variant="outlined">
                    <Avatar src="https://cdn.icon-icons.com/icons2/2643/PNG/512/male_boy_person_people_avatar_icon_159358.png" />
                  </List>
                  <List className="profile" variant="outlined">
                    <Span>{user.email}</Span>
                  </List>
                  <List variant="outlined" onClick={() => logout()}>
                    <ExitToAppIcon />
                    Logout
                  </List>
                </Profile>
              ) : null}
            </RelativeWrapper>
          </>
        ) : (
          <>
            <ActionButton
              variant="outlined"
              outline="outline"
              isTransparent={isTransparent}
              onClick={() => setState({ ...state, loginOpen: true })}
            >
              Login
            </ActionButton>
            <ActionButton
              variant="contained"
              outline="contained"
              isTransparent={isTransparent}
              onClick={() => setState({ ...state, open: true })}
            >
              SignUp
            </ActionButton>{" "}
          </>
        )}
        <Dialog
          open={open}
          onClose={() => setState({ ...state, open: false })}
          scroll="paper"
          aria-labelledby="scroll-dialog-title"
          aria-describedby="scroll-dialog-description"
          fullWidth
          maxWidth="sm"
        >
          <DialogContent>
            <Signup padding="2rem 2rem" isModal />
          </DialogContent>
        </Dialog>
        <Dialog
          open={loginOpen}
          onClose={() => setState({ ...state, loginOpen: false })}
          scroll="paper"
          aria-labelledby="scroll-dialog-title"
          aria-describedby="scroll-dialog-description"
          fullWidth
          maxWidth="sm"
        >
          <DialogContent>
            <Login
              padding="2rem 2rem"
              isModal
              handleDrawerClose={() => setState({ ...state, loginOpen: false })}
            />
          </DialogContent>
        </Dialog>

        <Dialog
          open={aboutOpen}
          onClose={() => setState({ ...state, aboutOpen: false })}
          scroll="paper"
          aria-labelledby="scroll-dialog-title"
          aria-describedby="scroll-dialog-description"
          fullWidth
          maxWidth="sm"
        >
          <DialogContent>
            <About />
          </DialogContent>
        </Dialog>
      </>
    );
  };

  const getDrawerChoices = () => {
    return (
      <>
        {headersData.map(({ label }) => {
          return <MenuItem key={label}>{label}</MenuItem>;
        })}
      </>
    );
  };
  // Desktop Nav
  const displayDesktop = () => {
    return (
      <StyledToolbar>
        {tourxLogo}
        <div>{getMenuButtons()}</div>
      </StyledToolbar>
    );
  };

  // Mobile View
  const displayMobile = () => {
    const handleDrawerOpen = () =>
      setState((prevState) => ({ ...prevState, drawerOpen: true }));
    const handleDrawerClose = () =>
      setState((prevState) => ({ ...prevState, drawerOpen: false }));

    return (
      <Toolbar>
        <MenuIconButton
          {...{
            edge: "start",
            color: "primary",
            "aria-label": "menu",
            "aria-haspopup": "true",
            onClick: handleDrawerOpen,
          }}
          isTransparent={isTransparent}
        >
          <MenuIcon />
        </MenuIconButton>

        <Drawer
          {...{
            anchor: "left",
            open: drawerOpen,
            onClose: handleDrawerClose,
          }}
        >
          <DrawerContainer>{getDrawerChoices()}</DrawerContainer>
        </Drawer>

        <div>{tourxLogo}</div>
        <EndWrapper>
          {isAuth ? (
            <>
              <ActionButton
                variant="outlined"
                outline="outline"
                isTransparent={isTransparent}
              >
                My Account
              </ActionButton>
            </>
          ) : (
            <>
              <ActionButton
                variant="outlined"
                outline="outline"
                isTransparent={isTransparent}
                onClick={() => setState({ ...state, loginOpen: true })}
              >
                Login
              </ActionButton>
              <ActionButton
                variant="contained"
                outline="contained"
                isTransparent={isTransparent}
                onClick={() => setState({ ...state, open: true })}
              >
                SignUp
              </ActionButton>
            </>
          )}

          <Dialog
            open={open}
            onClose={() => setState({ ...state, open: false })}
            scroll="paper"
            aria-labelledby="scroll-dialog-title"
            aria-describedby="scroll-dialog-description"
            fullWidth
            maxWidth="sm"
          >
            <DialogContent>
              <Signup padding="2rem 2rem" isModal />
            </DialogContent>
          </Dialog>
          <Dialog
            open={loginOpen}
            onClose={() => setState({ ...state, loginOpen: false })}
            scroll="paper"
            aria-labelledby="scroll-dialog-title"
            aria-describedby="scroll-dialog-description"
            fullWidth
            maxWidth="sm"
          >
            <DialogContent>
              <Login
                padding="2rem 2rem"
                isModal
                handleDrawerClose={() =>
                  setState({ ...state, loginOpen: false })
                }
              />
            </DialogContent>
          </Dialog>
        </EndWrapper>
      </Toolbar>
    );
  };

  return (
    <HeaderWrapper elevation={isTransparent ? "none" : elevation}>
      <header>
        <StyledAppBar color={isTransparent ? "transparent" : "inherit"}>
          {mobileView ? displayMobile() : displayDesktop()}
        </StyledAppBar>
      </header>
    </HeaderWrapper>
  );
};

Header.propTypes = {
  isTransparent: PropTypes.bool,
  elevation: PropTypes.string,
  isAuth: PropTypes.bool,
  setIsAuth: PropTypes.func,
  logoutUser: PropTypes.func,
  user: PropTypes.object,
};

Header.defaultProps = {
  isTransparent: false,
  elevation: "apply",
  isAuth: false,
  setIsAuth: () => {},
  logoutUser: () => {},
  user: {},
};

const mapStateToProps = (state) => ({
  isAuth: state.auth.isAuth,
  user: state.auth.user,
});

export default connect(mapStateToProps, { setIsAuth, logoutUser })(Header);
