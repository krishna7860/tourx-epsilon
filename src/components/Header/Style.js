/* eslint-disable no-nested-ternary */
/* eslint-disable import/prefer-default-export */
import styled from "styled-components";
import { Toolbar, Button, AppBar, IconButton } from "@material-ui/core";
import { Link } from "react-router-dom";

export const HeaderWrapper = styled.div`
  .MuiPaper-elevation4 {
    box-shadow: ${(props) =>
      props.elevation === "none"
        ? "none"
        : `0px 2px 4px -1px rgb(0 0 0 / 20%),
      0px 4px 5px 0px rgb(0 0 0 / 14%), 0px 1px 10px 0px rgb(0 0 0 / 12%)`} !important;
  }
`;

export const StyledToolbar = styled(Toolbar)`
  display: flex;
  justify-content: space-between;
`;

export const DrawerContainer = styled.div`
  padding: 20px 30px;
`;

export const LogoItem = styled(Link)`
  &:link {
    text-decoration: none;
  }
`;
export const InnerItem = styled.div`
  font-weight: 600;
  color: ${(props) => (props.isColored ? "#fffefe" : props.theme.colors.black)};
  text-align: left;
  font-size: 30px !important;
  @media screen and (max-width: 540px) {
    font-size: 24px !important;
  }
`;

export const MenuButton = styled(Button)`
  font-weight: 700;
  size: 18px;
  color: ${(props) =>
    props.isTransparent
      ? props.theme.colors.white
      : props.theme.colors.black} !important;
  margin-left: 38px;
  padding: 8px 20px !important;
`;
export const StyledAppBar = styled(AppBar)`
  background-color: #400ccc;
  padding-right: 118px;
  padding-left: 118px;
  padding-top: 10px;
  height: 80px;
  @media (max-width: 900px) {
    padding-left: 0px !important;
    padding-right: 0px;
  }
  @media screen and (max-width: 540px) {
    height: auto !important;
    padding-left: 0px !important;
    padding-right: 0px;
    padding-top: 5px;
    padding-bottom: 5px;
  }
`;

export const ActionButton = styled(Button)`
  background-color: ${(props) =>
    props.outline === "outline"
      ? "transparent"
      : props.theme.colors.black} !important ;
  color: ${(props) =>
    props.isTransparent && props.outline === "contained"
      ? props.theme.colors.white
      : !props.isTransparent && props.outline === "contained"
      ? props.theme.colors.white
      : props.isTransparent && props.outline === "outline"
      ? props.theme.colors.white
      : !props.isTransparent && props.outline === "outline"
      ? props.theme.colors.black
      : props.theme.colors.white} !important;
  border: ${(props) =>
    props.outline === "outline" ? "1px solid" : "none"} !important;
  border-color: ${(props) =>
    props.isTransparent ? "white" : "black"} !important;
  margin: 8px 20px !important;
  @media (max-width: 500px) {
    margin: 8px 5px !important;
  }
`;

export const MenuIconButton = styled(IconButton)`
  color: ${(props) => (props.isTransparent ? "white" : "black")} !important;
`;

export const EndWrapper = styled.div`
  display: flex;
  justify-content: flex-end;
  width: 100%;
`;

export const RelativeWrapper = styled.div`
  position: relative;
`;

export const Profile = styled.div`
  position: absolute;
  width: 220px;
  background: white;
  border: black;
  border-radius: 10px;
  left: 80px;
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);

  & > div.profile : {
    background: #eaecee;
  }
`;

export const List = styled.div`
  width: 100%;
  padding: 1rem 1.5rem;
  font-size: 18px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  border-radius: 10px;
  cursor: pointer;
  &.profile {
    justify-content: center;
    padding-bottom: 0.5rem;
  }
`;

export const Avatar = styled.img`
  height: 64px;
  width: 64px;
  border-radius: 50%;
`;

export const Span = styled.p`
  font-size: 14px;
`;
