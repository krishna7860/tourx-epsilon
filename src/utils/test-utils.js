/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable react/jsx-filename-extension */
/* eslint-disable react/prop-types */
import React from "react";
import { render as rtlRender } from "@testing-library/react";
// import { renderHook as rtlRenderHooks } from "@testing-library/react-hooks";
import { createStore } from "redux";
import { Provider } from "react-redux";
import { ThemeProvider } from "styled-components";
import * as themes from "../theme/schema.json";
import reducer from "../redux/rootReducer";
// import defaultStore from "../redux/defaultStore";

function render(ui, options = {}) {
  const {
    initialState = {},
    store = createStore(reducer, initialState),
    theme,
    ...renderOptions
  } = options;

  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <ThemeProvider theme={themes.data.light}>{children}</ThemeProvider>
      </Provider>
    );
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// function renderHookWithRedux(callback, options = {}) {
//   const { initialState = defaultStore, store = createStore(reducer, initialState), ...others } = options

//   const returned = rtlRenderHooks(callback, {
//     ...others,
//     // eslint-disable-next-line react/display-name
//     wrapper: ({ children }) => <Provider store={store}>{children}</Provider>
//   })

//   return {
//     ...returned,
//     store
//   }
// }

// re-export everything
export * from "@testing-library/react";
// override render method
export { render, rtlRender as renderHook };
