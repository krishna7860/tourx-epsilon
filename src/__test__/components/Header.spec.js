/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable react/jsx-filename-extension */
import React from "react";
import { MemoryRouter, Route } from "react-router";
import { render, cleanup } from "../../utils/test-utils";
import Header from "../../components/Header/Header";

describe("components/Header", () => {
  afterEach(cleanup);
  it("should be rendered properly", () => {
    const { asFragment } = render(
      <MemoryRouter initialEntries={[`/`]}>
        <Route path="/">
          <Header isTransparent={false} elevation="none" />
        </Route>
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
