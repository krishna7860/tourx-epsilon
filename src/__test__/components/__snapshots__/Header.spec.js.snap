// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Header should be rendered properly 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM WKGpf"
    elevation="none"
  >
    <header>
      <header
        class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorInherit sc-gKAaRy dvkorR mui-fixed MuiPaper-elevation4"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular sc-gtsrHT iddZFI MuiToolbar-gutters"
        >
          <a
            class="sc-hKFxyN btYNzt"
            href="/"
          >
            <div
              class="sc-eCApnc bCwWAa"
            >
              TourX
            </div>
          </a>
          <div>
            <a
              href="/category/9023132"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text sc-jSFjdj TSbCx"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Categories
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </a>
            <a
              href="/"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text sc-jSFjdj TSbCx"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  About us
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </a>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined sc-iCoGMd fXLviD"
              outline="outline"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Login
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained sc-iCoGMd bsVGjY"
              outline="contained"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                SignUp
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </header>
    </header>
  </div>
</DocumentFragment>
`;
